File: sim/src/core/city/presets.ts
/**
 * City presets + map sizes. These do NOT import real GIS data — they retune the
 * tensor-field generator so each city reads like its real counterpart: grid
 * regularity, downtown pull, coastline, and sprawl. Everything stays procedural
 * and seed-deterministic; a preset just picks the knobs.
 */

export type MapSize = 'small' | 'medium' | 'large';

/** World edge length in meters. Field cell size is fixed, so bigger = more cells. */
export const MAP_SIZE_METERS: Record<MapSize, number> = {
  small: 8000,
  medium: 12000, // the original default
  large: 18000,
};

export interface WaterConfig {
  /** a straight coastline (ocean / great lake) along one edge */
  coast: boolean;
  /** fixed coast bearing in degrees, or null for a seed-random bearing */
  coastAngleDeg: number | null;
  /** 0..1 — how far inland the coast sits (higher = more land) */
  coastInset: number;
  /** carve a meandering river */
  river: boolean;
}

export interface CityPreset {
  key: string;
  label: string;
  blurb: string;
  /** backed by a real OpenStreetMap import (real roads + coastline) */
  real?: boolean;
  /** street-grid regularity */
  grid: {
    /** tensor grid patch weight (higher = streets snap harder to the grid) */
    weight: number;
    /** base grid bearing in degrees; grids all align to it when rigid */
    angleDeg: number;
    /** true = rigid rectilinear (NYC/Chicago); false = organic (Boston) */
    rigid: boolean;
    /** field noise weight — the wobble in street direction */
    noiseWeight: number;
  };
  /** downtown radial convergence (Atlanta's highways vs a flat grid) */
  radialWeight: number;
  water: WaterConfig;
  /** >1 spreads density out (LA/Atlanta sprawl); <1 concentrates it (NYC) */
  sprawl: number;
}

const GENERIC: CityPreset = {
  key: 'generic',
  label: 'Random City',
  blurb: 'A fresh procedural city each seed.',
  grid: { weight: 1, angleDeg: 0, rigid: false, noiseWeight: 0.22 },
  radialWeight: 2.2,
  water: { coast: true, coastAngleDeg: null, coastInset: 0.7, river: true },
  sprawl: 1,
};

export const CITY_PRESETS: CityPreset[] = [
  GENERIC,
  {
    key: 'nyc',
    label: 'New York',
    real: true,
    blurb: 'Real Manhattan street grid between the Hudson and East River (OSM).',
    grid: { weight: 1.5, angleDeg: 29, rigid: true, noiseWeight: 0.06 },
    radialWeight: 1.4,
    water: { coast: true, coastAngleDeg: 120, coastInset: 0.78, river: true },
    sprawl: 0.72,
  },
  {
    key: 'chicago',
    label: 'Chicago',
    real: true,
    blurb: 'Real Chicago: the grid, Lake Michigan, and the branching river (OSM).',
    grid: { weight: 1.6, angleDeg: 0, rigid: true, noiseWeight: 0.05 },
    radialWeight: 1.6,
    water: { coast: true, coastAngleDeg: 0, coastInset: 0.82, river: true },
    sprawl: 0.95,
  },
  {
    key: 'la',
    label: 'Los Angeles',
    real: true,
    blurb: 'Real downtown LA: the sprawling grid and the LA River (OSM).',
    grid: { weight: 1.2, angleDeg: 12, rigid: true, noiseWeight: 0.12 },
    radialWeight: 0.9,
    water: { coast: true, coastAngleDeg: 210, coastInset: 0.85, river: false },
    sprawl: 1.7,
  },
  {
    key: 'boston',
    label: 'Boston',
    real: true,
    blurb: 'Real Boston: harbor, the Charles, and the downtown peninsula (OSM).',
    grid: { weight: 0.7, angleDeg: 40, rigid: false, noiseWeight: 0.5 },
    radialWeight: 2.6,
    water: { coast: true, coastAngleDeg: 75, coastInset: 0.62, river: true },
    sprawl: 0.85,
  },
  {
    key: 'atlanta',
    label: 'Atlanta',
    real: true,
    blurb: 'Real Atlanta: landlocked sprawl fanning out along the highways (OSM).',
    grid: { weight: 0.9, angleDeg: 20, rigid: false, noiseWeight: 0.3 },
    radialWeight: 3.2,
    water: { coast: false, coastAngleDeg: null, coastInset: 1, river: false },
    sprawl: 1.8,
  },
  {
    key: 'cleveland',
    label: 'Cleveland',
    real: true,
    blurb: 'Real Cleveland: Lake Erie and the winding Cuyahoga through the Flats (OSM).',
    grid: { weight: 1.3, angleDeg: 8, rigid: true, noiseWeight: 0.1 },
    radialWeight: 1.8,
    water: { coast: true, coastAngleDeg: 0, coastInset: 0.8, river: true },
    sprawl: 1.1,
  },
  {
    key: 'philly',
    label: 'Philadelphia',
    real: true,
    blurb: 'Real Center City: the William Penn grid between the Schuylkill and Delaware (OSM).',
    grid: { weight: 1.55, angleDeg: 0, rigid: true, noiseWeight: 0.06 },
    radialWeight: 1.5,
    water: { coast: false, coastAngleDeg: null, coastInset: 1, river: true },
    sprawl: 0.9,
  },
  {
    key: 'sf',
    label: 'San Francisco',
    real: true,
    blurb: 'Real San Francisco: hills, the bay, and a tight downtown grid (OSM).',
    grid: { weight: 1.35, angleDeg: 0, rigid: true, noiseWeight: 0.18 },
    radialWeight: 2.0,
    water: { coast: true, coastAngleDeg: 45, coastInset: 0.7, river: false },
    sprawl: 0.8,
  },
  {
    key: 'dc',
    label: 'Washington',
    real: true,
    blurb: 'Real Washington: L\'Enfant avenues, the Mall, and the Potomac (OSM).',
    grid: { weight: 1.1, angleDeg: 0, rigid: false, noiseWeight: 0.2 },
    radialWeight: 2.8,
    water: { coast: false, coastAngleDeg: null, coastInset: 1, river: true },
    sprawl: 1.05,
  },
  {
    key: 'seattle',
    label: 'Seattle',
    real: true,
    blurb: 'Real Seattle: Elliott Bay, Lake Union, and the downtown ridge (OSM).',
    grid: { weight: 1.25, angleDeg: 0, rigid: true, noiseWeight: 0.14 },
    radialWeight: 2.2,
    water: { coast: true, coastAngleDeg: 270, coastInset: 0.72, river: false },
    sprawl: 1.15,
  },
];

export function presetByKey(key: string | undefined): CityPreset {
  return CITY_PRESETS.find((p) => p.key === key) ?? GENERIC;
}

File: crates/mf-sim/src/city/presets.rs
//! City presets + map sizes. Port of `sim/src/core/city/presets.ts`.
//!
//! Presets do NOT import real GIS data; they retune the tensor-field generator
//! so each city reads like its real counterpart (grid regularity, downtown
//! pull, coastline, sprawl). Everything stays procedural + seed-deterministic;
//! a preset just picks the knobs.

/// Map size class. Mirrors `MapSize`.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum MapSize {
    /// 8 km world.
    Small,
    /// 12 km world (the default).
    Medium,
    /// 18 km world.
    Large,
}

impl MapSize {
    /// World edge length in meters. Mirrors `MAP_SIZE_METERS`.
    pub fn meters(self) -> f64 {
        match self {
            MapSize::Small => 8000.0,
            MapSize::Medium => 12000.0,
            MapSize::Large => 18000.0,
        }
    }
}

/// Water configuration for a preset. Mirrors `WaterConfig`.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct WaterConfig {
    /// A straight coastline (ocean / great lake) along one edge.
    pub coast: bool,
    /// Fixed coast bearing in degrees, or `None` for a seed-random bearing.
    pub coast_angle_deg: Option<f64>,
    /// 0..1 how far inland the coast sits (higher = more land).
    pub coast_inset: f64,
    /// Carve a meandering river.
    pub river: bool,
}

/// Street-grid regularity knobs. Mirrors `CityPreset.grid`.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct GridConfig {
    /// Tensor grid patch weight (higher = streets snap harder to the grid).
    pub weight: f64,
    /// Base grid bearing in degrees; grids all align to it when rigid.
    pub angle_deg: f64,
    /// `true` = rigid rectilinear (NYC/Chicago); `false` = organic (Boston).
    pub rigid: bool,
    /// Field noise weight (the wobble in street direction).
    pub noise_weight: f64,
}

/// A city preset. Mirrors `CityPreset`.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct CityPreset {
    /// Stable key.
    pub key: &'static str,
    /// Display label.
    pub label: &'static str,
    /// Backed by a real OpenStreetMap import (real roads + coastline).
    pub real: bool,
    /// Street-grid regularity.
    pub grid: GridConfig,
    /// Downtown radial convergence.
    pub radial_weight: f64,
    /// Water configuration.
    pub water: WaterConfig,
    /// >1 spreads density out (sprawl); <1 concentrates it.
    pub sprawl: f64,
}

const GENERIC: CityPreset = CityPreset {
    key: "generic",
    label: "Random City",
    real: false,
    grid: GridConfig {
        weight: 1.0,
        angle_deg: 0.0,
        rigid: false,
        noise_weight: 0.22,
    },
    radial_weight: 2.2,
    water: WaterConfig {
        coast: true,
        coast_angle_deg: None,
        coast_inset: 0.7,
        river: true,
    },
    sprawl: 1.0,
};

/// All city presets, in the same order as `CITY_PRESETS`.
pub const CITY_PRESETS: &[CityPreset] = &[
    GENERIC,
    CityPreset {
        key: "nyc",
        label: "New York",
        real: true,
        grid: GridConfig {
            weight: 1.5,
            angle_deg: 29.0,
            rigid: true,
            noise_weight: 0.06,
        },
        radial_weight: 1.4,
        water: WaterConfig {
            coast: true,
            coast_angle_deg: Some(120.0),
            coast_inset: 0.78,
            river: true,
        },
        sprawl: 0.72,
    },
    CityPreset {
        key: "chicago",
        label: "Chicago",
        real: true,
        grid: GridConfig {
            weight: 1.6,
            angle_deg: 0.0,
            rigid: true,
            noise_weight: 0.05,
        },
        radial_weight: 1.6,
        water: WaterConfig {
            coast: true,
            coast_angle_deg: Some(0.0),
            coast_inset: 0.82,
            river: true,
        },
        sprawl: 0.95,
    },
    CityPreset {
        key: "la",
        label: "Los Angeles",
        real: true,
        grid: GridConfig {
            weight: 1.2,
            angle_deg: 12.0,
            rigid: true,
            noise_weight: 0.12,
        },
        radial_weight: 0.9,
        water: WaterConfig {
            coast: true,
            coast_angle_deg: Some(210.0),
            coast_inset: 0.85,
            river: false,
        },
        sprawl: 1.7,
    },
    CityPreset {
        key: "boston",
        label: "Boston",
        real: true,
        grid: GridConfig {
            weight: 0.7,
            angle_deg: 40.0,
            rigid: false,
            noise_weight: 0.5,
        },
        radial_weight: 2.6,
        water: WaterConfig {
            coast: true,
            coast_angle_deg: Some(75.0),
            coast_inset: 0.62,
            river: true,
        },
        sprawl: 0.85,
    },
    CityPreset {
        key: "atlanta",
        label: "Atlanta",
        real: true,
        grid: GridConfig {
            weight: 0.9,
            angle_deg: 20.0,
            rigid: false,
            noise_weight: 0.3,
        },
        radial_weight: 3.2,
        water: WaterConfig {
            coast: false,
            coast_angle_deg: None,
            coast_inset: 1.0,
            river: false,
        },
        sprawl: 1.8,
    },
    CityPreset {
        key: "cleveland",
        label: "Cleveland",
        real: true,
        grid: GridConfig {
            weight: 1.3,
            angle_deg: 8.0,
            rigid: true,
            noise_weight: 0.1,
        },
        radial_weight: 1.8,
        water: WaterConfig {
            coast: true,
            coast_angle_deg: Some(0.0),
            coast_inset: 0.8,
            river: true,
        },
        sprawl: 1.1,
    },
    CityPreset {
        key: "philly",
        label: "Philadelphia",
        real: true,
        grid: GridConfig {
            weight: 1.55,
            angle_deg: 0.0,
            rigid: true,
            noise_weight: 0.06,
        },
        radial_weight: 1.5,
        water: WaterConfig {
            coast: false,
            coast_angle_deg: None,
            coast_inset: 1.0,
            river: true,
        },
        sprawl: 0.9,
    },
    CityPreset {
        key: "sf",
        label: "San Francisco",
        real: true,
        grid: GridConfig {
            weight: 1.35,
            angle_deg: 0.0,
            rigid: true,
            noise_weight: 0.18,
        },
        radial_weight: 2.0,
        water: WaterConfig {
            coast: true,
            coast_angle_deg: Some(45.0),
            coast_inset: 0.7,
            river: false,
        },
        sprawl: 0.8,
    },
    CityPreset {
        key: "dc",
        label: "Washington",
        real: true,
        grid: GridConfig {
            weight: 1.1,
            angle_deg: 0.0,
            rigid: false,
            noise_weight: 0.2,
        },
        radial_weight: 2.8,
        water: WaterConfig {
            coast: false,
            coast_angle_deg: None,
            coast_inset: 1.0,
            river: true,
        },
        sprawl: 1.05,
    },
    CityPreset {
        key: "seattle",
        label: "Seattle",
        real: true,
        grid: GridConfig {
            weight: 1.25,
            angle_deg: 0.0,
            rigid: true,
            noise_weight: 0.14,
        },
        radial_weight: 2.2,
        water: WaterConfig {
            coast: true,
            coast_angle_deg: Some(270.0),
            coast_inset: 0.72,
            river: false,
        },
        sprawl: 1.15,
    },
];

/// The `generic` preset. Mirrors the default `GENERIC`.
pub fn generic() -> CityPreset {
    GENERIC
}

/// Look up a preset by key, falling back to `generic`. Mirrors `presetByKey`.
pub fn preset_by_key(key: Option<&str>) -> CityPreset {
    match key {
        Some(k) => CITY_PRESETS
            .iter()
            .find(|p| p.key == k)
            .copied()
            .unwrap_or(GENERIC),
        None => GENERIC,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn lookup_hits_and_falls_back() {
        assert_eq!(preset_by_key(Some("nyc")).label, "New York");
        assert_eq!(preset_by_key(Some("nope")).key, "generic");
        assert_eq!(preset_by_key(None).key, "generic");
    }

    #[test]
    fn map_sizes() {
        assert_eq!(MapSize::Medium.meters(), 12000.0);
    }
}
